pkg/i2gw/provider.go

```diff
@@ -99,14 +108,106 @@ type GatewayResources struct {
 	UDPRoutes  map[types.NamespacedName]gatewayv1alpha2.UDPRoute
 
 	ReferenceGrants map[types.NamespacedName]gatewayv1beta1.ReferenceGrant
+
+	GatewayExtensions map[types.NamespacedName]unstructured.Unstructured
+}
+
+// IR holds specifications of Gateway Objects for supporting Ingress extensions,
+// annotations, and proprietary API features not supported as Gateway core
+// features. An IR field can be mapped to core Gateway-API fields,
+// or provider-specific Gateway extensions.
+type IR struct {
+	Gateways   map[types.NamespacedName]GatewayContext
+	HTTPRoutes map[types.NamespacedName]HTTPRouteContext
+	Services   map[types.NamespacedName]*ServiceIR
+
+	GatewayClasses map[types.NamespacedName]gatewayv1.GatewayClass
+	TLSRoutes      map[types.NamespacedName]gatewayv1alpha2.TLSRoute
+	TCPRoutes      map[types.NamespacedName]gatewayv1alpha2.TCPRoute
+	UDPRoutes      map[types.NamespacedName]gatewayv1alpha2.UDPRoute
+
+	ReferenceGrants map[types.NamespacedName]gatewayv1beta1.ReferenceGrant
+}
+
+// GatewayContext contains the Gateway-API Gateway object, along with a
+// dedicated field for each provider to specify their extension features on
+// Gateway.
+// The IR will contain necessary information to construct the Gateway
+// extensions, but not the extensions themselves.
+type GatewayContext struct {
+	gatewayv1.Gateway
+
+	Apisix       *ApisixGatewayIR
+	GCE          *GceGatewayIR
+	IngressNginx *IngressNginxGatewayIR
+	Istio        *IstioGatewayIR
+	Kong         *KongGatewayIR
+	Openapi3     *Openapi3GatewayIR
```

Comment:
Thinking out loud - can we somehow make each provider register itself and his IR type with the GatewayContext? similarly to how we did with ProviderConstructorByName [1](https://github.com/kubernetes-sigs/ingress2gateway/blob/c626d145e4eae55a6ced3fd204afee9f676eace0/pkg/i2gw/provider.go#L35C5-L35C30 ) & [2](https://github.com/kubernetes-sigs/ingress2gateway/blob/c626d145e4eae55a6ced3fd204afee9f676eace0/pkg/i2gw/providers/apisix/apisix.go#L31)?

If possible, it is going to be easier to maintain and a bit cleaner

Reply:
Yes that's a good suggestion! I think we can make them into a map of string to interface{}, so each provider can implement the IR however they wanted.

Reply:
Updated, please take a look!